Introduce new PSR-18 based client

Art4 opened the pull request:
Hey,

I would like to introduce a HTTP client agnostic client based on [PSR-18](https://www.php-fig.org/psr/psr-18/) and [PSR-17](https://www.php-fig.org/psr/psr-17/).

I'm not done yet, but would like to hear what you have to say about it.

It is important to me that all changes in this PR are backward compatible. If you find a BC break I would love to fix ist. But maybe this new client could also be a step to a v2 of this project.

## Benefits of this PR

1. Allows the client consuming an app-wide PSR-18 HTTP client (like Guzzle :tada:)
2. Makes the client a lot more simpler, because all curl opt handling can be removed
3. No need for port handling, ssl checks and host customization
4. Allows custom API implementations for not implemented APIs

## Switching to `Psr18Client`

As I said before all changes in this PR are backward compatible. However switching to the new Psr18Client will create some work on the client side. 

1. Requires a PSR-18 HTTP client (like Guzzle).
2. Requires a PSR-17 `ServerRequestFactory` and ...
3. ... a `StreamFactory`. This can be done with an anonymous class using `guzzlehttp/psr7`
4. All curls options must be set to the PSR-18 client (if it is a curl based client) or to the PSR-7 Request (can be done with a middleware)

As you can see in the todo list im planning to create an "migrate guide" to explain the alternatives for features that will not exists in the `Psr18Client` (like port handling, ssl checks and host customization).

I also have to determine what this PR will mean for an installation without `composer`.

## Todo

- [x] Create Client interface
- [x] Create PSR-18 consuming client
- [x] Move handling of JSON and XML decoding outside of client
- [x] Make Api implementation client agnostic
- [x] Implement Redmine Authentication
- [x] Implement Redmine User impersonation
- [x] Add tests for json and xml decoding
- [x] Add tests for file uploads 
- [x] Update `example.php`
- [x] Check installation without composer (if necessary)

Cheers 🥂 

Update: Creating the migrate guide will be made in another PR.

Art4 commented:
I've checked the possibilities for using this library without composer. Because new dependency to `psr/http-client` and `psr/http-factory` will been created, the way currently described will no longer work.

Fortunately, there are now online tools that can be used instead of doing this manually:
https://php-download.com/package/kbsali/redmine-api

I think, this should allow us to remove the `src/autoload.php` and simply refer to php-download.com

kbsali commented:
cool @Art4 , I didn't know about this service! :+1:

kbsali commented on README.md:
```diff
@@ -45,11 +45,11 @@ A possible solution to this would be to create an extra APIs implementing the mi
 ## Requirements
 
-* PHP >= 5.4
+* PHP >= 7.3 or PHP 8.0
```
```suggestion
* PHP >= 7.3
```

Art4 replied:
I wanted to make clear that this library will be 8.0 ready. What about composer-like statement? `PHP ^7.3 || ^8.0`

kbsali replied:
sure

kbsali commented on README.md:
````diff
@@ -61,7 +61,7 @@ A possible solution to this would be to create an extra APIs implementing the mi
 
 ```bash
-$ php composer.phar require kbsali/redmine-api:~1.0
+$ php composer.phar require kbsali/redmine-api
````
```suggestion
$ composer require kbsali/redmine-api
```
? the .phar is the very old way of playing with composer, don't you think?

Art4 replied:
I'm still using the `composer.phar` in almost all projects to not have a globally dependency. :blush:
I would instead look at the possible users of the library. For advanced user composer will be clear. But in the README.php and example.php it suggests that this library aims also beginners.

kbsali commented on README.md:
````diff
@@ -110,19 +110,22 @@ You can run test suite to make sure the library will work properly on your syste
 ```
 $ vendor/bin/phpunit
-PHPUnit 6.5.14 by Sebastian Bergmann and contributors.
+PHPUnit 9.5.3 by Sebastian Bergmann and contributors.
````
:heart_eyes:

kbsali commented on docs/usage.md:
```diff
@@ -0,0 +1,500 @@
+# Usage
+
+:warning: This file shows example commands for reading, creating, updating and deleting redmine components. As this are only examples, please **make sure, that you do not execute them at your production Redmine instance and may destroy real life content!**
```
```suggestion
:warning: This file shows example commands for reading, creating, updating and deleting Redmine data. As these are only examples, please **make sure, that you do not execute them against your production instance as it may break real content!**
```

kbsali commented on docs/usage.md:
````diff
@@ -0,0 +1,500 @@
+Please make sure you have install `php-remine-api` with an PSR-4 autoloader like [Composer](https://getcomposer.org/).
+
+```bash
+php composer.phar require kbsali/php-redmine-api
````
```suggestion
composer require kbsali/php-redmine-api
```

kbsali commented on docs/usage.md:
````diff
@@ -0,0 +1,500 @@
++$client = new Redmine\Client('http://redmine.example.com', 'username', 'password');
+```
+
+> :bulb: For security reason it is recommended that you use a ApiKey instead of username/password.
````
```suggestion
> :bulb: For security reason it is recommended that you use an ApiKey rather than your username/password.
```

Art4 replied:
Thank you very much. :heart:  I'm not a native speaker.

kbsali commented on docs/usage.md:
```diff
@@ -0,0 +1,500 @@
+- an Apikey or username
+- and optional a password if you want tu use username/password.
+
+> :bulb: For security reason it is recommended that you use a ApiKey instead of username/password.
```
```suggestion
> :bulb: For security reason it is recommended that you use an ApiKey rather than your username/password.
```

kbsali commented on docs/usage.md:
```diff
@@ -0,0 +1,500 @@
+
+Because the `Psr18Client` is agnostic about the HTTP client implementation every configuration specific to the transport has to be set to the `Psr\Http\Client\ClientInterface` implementation.
+
+This means if you want to set any curl settings to guzzle you have multiple ways to set them:
```
```suggestion
This means that if you want to set any `cURL` settings to `Guzzle` you have multiple ways to set them:
```

kbsali commented on src/Redmine/Api/AbstractApi.php:
```diff
@@ -43,7 +46,27 @@ public function lastCallFailed()
+        if ($body !== '') {
+            // if response is XML, return an SimpleXMLElement object
+            if (0 === strpos($this->client->getLastResponseContentType(), 'application/xml')) {
+                $body = new SimpleXMLElement($body);
```
how about early return ?

Art4 replied:
Sure, good catch.

kbsali commented on src/Redmine/Api/AbstractApi.php:
```diff
@@ -43,7 +46,27 @@ public function lastCallFailed()
+
+        $body = $this->client->getLastResponseBody();
+
+        if ($body !== '') {
```
```suggestion
        if ($body !== '' && 0 === strpos($this->client->getLastResponseContentType(), 'application/xml')) {
            return new SimpleXMLElement($body);
```
?

kbsali commented on src/Redmine/Api/AbstractApi.php:
```diff
@@ -43,7 +46,27 @@ public function lastCallFailed()
+            }
+        }
+
+        if ($decode === true && $body !== '' and 0 === strpos($this->client->getLastResponseContentType(), 'application/json')) {
```
```suggestion
        if ($decode === true && $body !== '' && 0 === strpos($this->client->getLastResponseContentType(), 'application/json')) {
```

kbsali commented on src/Redmine/Api/AbstractApi.php:
```diff
@@ -43,7 +46,27 @@ public function lastCallFailed()
+
+        if ($decode === true && $body !== '' and 0 === strpos($this->client->getLastResponseContentType(), 'application/json')) {
+            try {
+                $body = json_decode($body, true, 512, \JSON_THROW_ON_ERROR);
```
```suggestion
                return json_decode($body, true, 512, \JSON_THROW_ON_ERROR);
```

kbsali commented on src/Redmine/Api/AbstractApi.php:
```diff
@@ -43,7 +46,27 @@ public function lastCallFailed()
+            try {
+                $body = json_decode($body, true, 512, \JSON_THROW_ON_ERROR);
+            } catch (JsonException $e) {
+                $body = 'Error decoding body as JSON: '.$e->getMessage();
```
```suggestion
                return 'Error decoding body as JSON: '.$e->getMessage();
```

kbsali commented on src/Redmine/Api/AbstractApi.php:
```diff
@@ -56,7 +79,18 @@ protected function get($path, $decode = true)
+
+        $body = $this->client->getLastResponseBody();
+
+        if ($body !== '') {
```
same proposal as above

kbsali commented on src/Redmine/Client.php:
```diff
@@ -105,6 +105,16 @@ class Client
+    /**
+     * @var string Redmine response content type
+     */
+    private $responseContentType = '';
```
```suggestion
    private string $responseContentType = '';
```
how about trying to type every (new) property?

Art4 replied:
This would be require PHP 7.4, but I tried to match PHP 7.3 as well.

kbsali commented on src/Redmine/Client.php:
```diff
@@ -204,6 +182,94 @@ public function getUrl()
+    /**
+     * HTTP GETs a json $path.
+     *
+     * @param string $path
```
i think we can remove all those docblocks when the method signature forces the input type hint and well as the return type

Art4 replied:
Sure, will do this. :+1:

kbsali commented on src/Redmine/Client.php:
```diff
@@ -204,6 +182,94 @@ public function getUrl()
     }
 
+    /**
+     * HTTP GETs a json $path.
```
too obvious to be mentioned, don't you think :)

Art4 replied:
I removed all obvious docblocks and rephrase the descriptions. I would like to keep them for consistency.

kbsali commented on src/Redmine/Client/ClientApiTrait.php:
```diff
@@ -0,0 +1,65 @@
+use Redmine\Api;
+
+/**
+ * client api trait.
```
not very helpful, i would remove it

Art4 replied:
I rephrased it and add an `@internal` docblock. What do you think?

kbsali commented on src/Redmine/Client/ClientApiTrait.php:
```diff
@@ -0,0 +1,65 @@
+    /**
+     * @var array APIs
+     */
+    private $apiInstances = [];
```
```suggestion
    private array $apiInstances = [];
```

kbsali commented on src/Redmine/Client/ClientApiTrait.php:
```diff
@@ -0,0 +1,65 @@
+     */
+    private $apiInstances = [];
+
+    private $apiClassnames = [
```
```suggestion
    private array $apiClassnames = [
```

kbsali reviewed (approved):
too long to review till the end! :sweat_smile: 
amazing work!!! :muscle: 
I would try to type hint everything (new at least) and get rid of most of the useless docblocks

But this can be done in another PR, up to you!

Art4 commented:
> I would try to type hint everything (new at least) and get rid of most of the useless docblocks
> 
> But this can be done in another PR, up to you!

Typed properties where released with PHP 7.4. I wanted to keep the PHP 7.3 compatibility. I would leave typed properties to a future release.

Art4 commented:
Thank you for your first review. I will admit your suggestions as best as possible. While working on this PR I developed this possible roadmap:

**Release 1.7**

- Introduce new `Redmine\Client\Psr18Client`, keep everything else as it is.
- This will allow testing the new client in the wild.

**Release 1.8**

- Bump required PHP version to 7.4, add typed properties
- Create new `Redmine\Client\NativeCurlClient` that will introduce new curl options setter and forbid/remove some old features.
- Let `Redmine\Client` extends from NativeCurlClient
- Add deprecation warnings in `Redmine\Client` for a smoother upgrade experience.

**Release 2.0**
- **breaking** Remove `Redmine\Client` 
- **breaking** Move `Redmine\Client\Client` interface to `Redmine\Client`
- **breaking** Make NativeCurlClient final

kbsali commented:
I like your plan! :)
one thing though: you are removing the `$client->issue` in favor of `$client->get(issue')`, I haven't tested... does it still work the "old way"? If not, this is quite a big breaking change :warning:

kbsali commented:
Also, why adding a new "usage.md" file? How about updating the README instead?

And final point : would you be interested in maintaining the package yourself?

Art4 commented:
> I like your plan! :)

Thank you. :+1: 

> one thing though: you are removing the `$client->issue` in favor of `$client->get(issue')`, I haven't tested... does it still work the "old way"? If not, this is quite a big breaking change warning

I introduced the method `getApi()` in the client interface. The old client still supports `$client->issue` and `$client->get(issue')`, but in `Psr18Client` one has to switch to `getApi()`.

I now also added tests to guaranty the old behavior on the old client.

> Also, why adding a new "usage.md" file? How about updating the README instead?

I thought about moving the example.php and examples from README into a new docs folder. There will be also the migration guide. This would blow up the README.md

But if you wish I could leave that in the README.md. Or we move this docs discussion into a separate PR.

> And final point : would you be interested in maintaining the package yourself?

It is a honor for me that you consider me as a maintainer, but I'm afraid I have to decline. I'm already maintaining some projects myself. But I'd be happy to contribute code and bugfixes and reply to issues if I find some time (like in the last days).

Art4 commented:
> Also, why adding a new "usage.md" file? How about updating the README instead?

I moved the usage docs into README.md. I will create the migration guide into another PR.

From my side this PR is ready for merge.

kbsali commented:
awesome work! thanks a lot!
(and too bad for the maintaining offer, if you change your mind please do let me know! :))
